src/api-binder.ts

```diff
@@ -0,0 +1,890 @@
+			);
+		}
+
+		try {
+			const deviceId = await this.config.get('deviceId');
+			const release = await this.balenaApi.get({
+				resource: 'release',
+				options: {
+					$filter: {
+						belongs_to__application: app,
+						commit,
+						status: 'success',
+					},
+					$select: 'id',
+				},
+			});
+
+			const releaseId = _.get(release, '[0].id');
+			if (releaseId == null) {
+				throw new Error(
+					'Cannot continue pinning preloaded device! No release found!',
+				);
+			}
+
+			await this.balenaApi.patch({
+				resource: 'device',
+				id: deviceId as number,
+				body: {
+					should_be_running__release: releaseId,
+				},
+			});
+
+			// Set the config value for pinDevice to null, so that we know the
+			// task has been completed
+			await this.config.remove('pinDevice');
+		} catch (e) {
+			console.log('Could not pin device to release!');
+			console.log('Error: ', e);
+			throw e;
+		}
+	}
+
+	// Creates the necessary config vars in the API to match the current device state,
+	// without overwriting any variables that are already set.
+	private async reportInitialEnv(apiEndpoint: string) {
+		if (this.balenaApi == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to report initial environment without an API client',
+			);
+		}
+
+		const targetConfigUnformatted = _.get(
+			await this.getTargetState(),
+			'local.config',
+		);
+		if (targetConfigUnformatted == null) {
+			throw new InternalInconsistencyError(
+				'Attempt to report initial state with malformed target state',
+			);
+		}
+
+		const defaultConfig = this.deviceState.deviceConfig.getDefaults();
+
+		const currentState = await this.deviceState.getCurrentForComparison();
+		const targetConfig = await this.deviceState.deviceConfig.formatConfigKeys(
+			targetConfigUnformatted,
+		);
+		const deviceId = await this.config.get('deviceId');
+
+		const currentConfig: Dictionary<string> = currentState.local.config;
+		for (const [key, value] of _.toPairs(currentConfig)) {
+			let varValue = value;
+			// We want to disable local mode when joining a cloud
+			if (key === 'SUPERVISOR_LOCAL_MODE') {
+				varValue = 'false';
+			}
+			// We never want to disable VPN if, for instance, it failed to start so far
+			if (key === 'SUPERVISOR_VPN_CONTROL') {
+				varValue = 'true';
+			}
+
+			if (targetConfig[key] == null && value !== defaultConfig[key]) {
+				const envVar = {
+					varValue,
+					device: deviceId,
+					name: 'RESIN_' + key,
+				};
+				await this.balenaApi.post({
+					resource: 'device_config_variable',
+					body: envVar,
+				});
+			}
+		}
+
+		await this.config.set({ initialConfigReported: apiEndpoint });
+	}
+
+	private async reportInitialConfig(
+		apiEndpoint: string,
+		retryDelay: number,
+	): Promise<void> {
+		try {
+			await this.reportInitialEnv(apiEndpoint);
+		} catch (err) {
+			console.error('Error reporting initial configuration, will retry', err);
+			return Bluebird.delay(retryDelay).then(() =>
```

Comment:
This can be
```suggestion
			await Bluebird.delay(retryDelay);
```
to keep the style more consistent